dissect/target/plugin.py

```diff
@@ -796,3 +824,113 @@ def __init_subclass__(cls, **kwargs):
+        yield from plugins()
+        yield from os_plugins()
+        yield from child_plugins()  # Although nothing is exported here...
+
+    for available in all_plugins():
+
+        modulepath = available["module"]
+        if modulepath.endswith("._os"):
+            modulepath = "OS"
+            prefix = ""
+        else:
+            prefix = f"{modulepath}."
+
+        rootset.add(modulepath.split(".")[0])
+        for exported in available["exports"]:
+            if exported == "get_all_records":
+                continue
+            index[f"{prefix}{exported}"] = available
+
+    return index, rootset
+
+
+def find_plugin_functions(target: Target, patterns: str, compatibility: bool = False) -> list[PluginFunction]:
+    result = []
+    functions, rootset = plugin_function_index(target)
+
+    for pattern in patterns.split(","):
+
+        wildcard = any(char in pattern for char in ["*", "!", "?", "[", "]"])
+        treematch = pattern.split(".")[0] in rootset and pattern != "os"
+
+        if treematch and not wildcard:
+            pattern += ".*"  # this seems to be what people expect...
+
+        if wildcard or treematch:
+            for index_name, func in functions.items():
+                if fnmatch.fnmatch(index_name, pattern):
+                    method_name = index_name.split(".")[-1]
+                    loaded_plugin_object = load(func)
+
+                    # some plugins don't want to be found when using wildcards...
+                    if not loaded_plugin_object.findable:
+                        continue
+
+                    fobject = inspect.getattr_static(loaded_plugin_object, method_name)
+
+                    if compatibility:
+                        try:
+                            if not loaded_plugin_object(target).is_compatible():
+                                continue
+                        except Exception:
+                            continue
+
+                    result.append(
+                        PluginFunction(
+                            name=index_name,
+                            class_object=loaded_plugin_object,
+                            method_name=method_name,
+                            output_type=getattr(fobject, "__output__", "text"),
+                            plugin_desc=func,
+                        )
+                    )
+        else:
+            # otherwise match using ~ classic style
+            if pattern.find(".") > -1:
+                namespace, funcname = pattern.split(".", 1)
+            else:
+                funcname = pattern
+                namespace = None
+
+            plugindesc = None
+            for index_name, func in functions.items():
+                nsmatch = namespace and func["namespace"] == namespace and funcname in func["exports"]
+                fmatch = not namespace and not func["namespace"] and funcname in func["exports"]
+                if nsmatch or fmatch:
+                    plugindesc = func
```

Comment:
Yes, I suggested this change on purpose since `getattr` might execute plugin code prematurely (before we establish based on static analysis of the return type whether we want to execute this function or not). Also the return value is dynamic which might lead to insecurities in the CLI (in theory you can never predict the effect unless we peparse the underlying Python code).